### ISSUE TITLE
Getting list items with LoadListDataAsStreamAsync with CAML query won't allow to request RoleAssignments

### ISSUE THREAD

**Issue by pxmonti** [pnp/pnpcore#607]
Hello all, I need to load all items from lists/libraries with potentially a large number of items and for this I thought I'd use a CAML query approach to extract only the fields I need and use **RowLimit** for pagination. Since I need also _*system*_ fields like **FieldRef** and **File_x0020_Size**, I cannot use **LoadItemsByCamlQueryAsync**, because as per documentation this method won't return those.
**LoadListDataAsStreamAsync** would solve this, however I also need to fetch the **RoleAssigments** collection for every item, but this method won't return it (I checked that **LoadItemsByCamlQueryAsync** would return **RoleAssigments** correctly loaded instead).
So, basically looks like there's no way as of now (that I am awar of) to query items using a CAML query and have system fields fetched and have **RoleAssigment** fetched all with a single call.
I was able to do so with the _*old*_ CSOM approach, so I was expecting I could do this with **PnP.Core SDK** too.
Is there a way to get to the result I'm trying to achieve?
Thank you

**Comment by jansenbe** [comment]
Thanks for using @pxmonti and providing feedback.

For now this indeed is the case, I'll do some more research to understand if somehow the LoadItemsByCamlQuery approach can return the "system" fields...if not then I'm considering adding a another data load option that under the covers uses the CSOM endpoint.

**Comment by pxmonti** [comment]
Thank you @jansenbe. For the time being I'll stick with the current CSOM approach and wait for news from your side.

**Comment by jansenbe** [comment]
That makes sense @pxmonti. I'll add a new issue to track the need

**Comment by jansenbe** [comment]
@pxmonti : can you share the CSOM snippet you're using today? Will help with ensuring we verify your use case.

**Comment by pxmonti** [comment]
Sure, this is what I currently do with CSOM:

``

          CamlQuery camlQuery = new CamlQuery();
          camlQuery.SetViewAttribute(QueryScope.RecursiveAll);
          camlQuery.SetViewFields(new string[] { "ID",
					          "Title",
					          "FileRef",
					          "Author",
					          "Editor",
					          "Created",
					          "Modified",
					          "File_x0020_Size"});
            camlQuery.SetQueryRowlimit(1000u);
            ListItemCollectionPosition position = null;
            do
            {
	            camlQuery.ListItemCollectionPosition = position;
	            ListItemCollection listItems = spList.GetItems(camlQuery);
	            context.Load(listItems, lits => lits.Include(
		            item => item.DisplayName,
		            item => item.HasUniqueRoleAssignments,
		            item => item.FileSystemObjectType,
		            item => item.Id,
		            item => item.RoleAssignments)
		            );
            
	            try
	            {
		            context.ExecuteQueryRetry(10, 500, decoration);
	            }
	            catch (Exception ex)
	            {
		            // Logging exceptions
	            }
                position = listItems.ListItemCollectionPosition;
	            /*
	            ... Do something with the items
	            */
            } while( position != null);

``

**Comment by jansenbe** [comment]
Adding some research notes here: using $select and $expand kind of works in combination with the `GetItems` call: below request 

Request: `https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/GetItems?$select=*,FileDirRef,File_x0020_Type,HasUniqueRoleAssignments&$expand=RoleAssignments`

Body:
```json
{
    "query": {
        "__metadata": {
            "type": "SP.CamlQuery"
        },
        "ViewXml": "<View><ViewFields><FieldRef Name='Title' /><FieldRef Name='FileDirRef' /><FieldRef Name='File_x0020_Type' /></ViewFields><OrderBy Override='TRUE'><FieldRef Name= 'ID' Ascending= 'FALSE' /></OrderBy></View>",
        "DatesInUtc": true
    }
}
```

Results in an expanded `RoleAssignments` + "system" fields like  `FileDirRef` and `File_x0020_Type`:

```json
{
    "d": {
        "results": [
            {
                "__metadata": {
                    "id": "3d5d7814-a5c7-4f66-9505-ba9f7bad0c41",
                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)",
                    "etag": "\"107\"",
                    "type": "SP.Data.FieldTypesListItem"
                },
                "FirstUniqueAncestorSecurableObject": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/FirstUniqueAncestorSecurableObject"
                    }
                },
                "RoleAssignments": {
                    "results": [
                        {
                            "__metadata": {
                                "id": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(3)",
                                "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(3)",
                                "type": "SP.RoleAssignment"
                            },
                            "Member": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(3)/Member"
                                }
                            },
                            "RoleDefinitionBindings": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(3)/RoleDefinitionBindings"
                                }
                            },
                            "PrincipalId": 3
                        },
                        {
                            "__metadata": {
                                "id": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(4)",
                                "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(4)",
                                "type": "SP.RoleAssignment"
                            },
                            "Member": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(4)/Member"
                                }
                            },
                            "RoleDefinitionBindings": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(4)/RoleDefinitionBindings"
                                }
                            },
                            "PrincipalId": 4
                        },
                        {
                            "__metadata": {
                                "id": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(5)",
                                "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(5)",
                                "type": "SP.RoleAssignment"
                            },
                            "Member": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(5)/Member"
                                }
                            },
                            "RoleDefinitionBindings": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(5)/RoleDefinitionBindings"
                                }
                            },
                            "PrincipalId": 5
                        },
                        {
                            "__metadata": {
                                "id": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(6)",
                                "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(6)",
                                "type": "SP.RoleAssignment"
                            },
                            "Member": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(6)/Member"
                                }
                            },
                            "RoleDefinitionBindings": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(6)/RoleDefinitionBindings"
                                }
                            },
                            "PrincipalId": 6
                        },
                        {
                            "__metadata": {
                                "id": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(16)",
                                "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(16)",
                                "type": "SP.RoleAssignment"
                            },
                            "Member": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(16)/Member"
                                }
                            },
                            "RoleDefinitionBindings": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(16)/RoleDefinitionBindings"
                                }
                            },
                            "PrincipalId": 16
                        },
                        {
                            "__metadata": {
                                "id": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(44)",
                                "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(44)",
                                "type": "SP.RoleAssignment"
                            },
                            "Member": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(44)/Member"
                                }
                            },
                            "RoleDefinitionBindings": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(44)/RoleDefinitionBindings"
                                }
                            },
                            "PrincipalId": 44
                        },
                        {
                            "__metadata": {
                                "id": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(47)",
                                "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(47)",
                                "type": "SP.RoleAssignment"
                            },
                            "Member": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(47)/Member"
                                }
                            },
                            "RoleDefinitionBindings": {
                                "__deferred": {
                                    "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/RoleAssignments/GetByPrincipalId(47)/RoleDefinitionBindings"
                                }
                            },
                            "PrincipalId": 47
                        }
                    ]
                },
                "AttachmentFiles": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/AttachmentFiles"
                    }
                },
                "ContentType": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/ContentType"
                    }
                },
                "GetDlpPolicyTip": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/GetDlpPolicyTip"
                    }
                },
                "FieldValuesAsHtml": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/FieldValuesAsHtml"
                    }
                },
                "FieldValuesAsText": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/FieldValuesAsText"
                    }
                },
                "FieldValuesForEdit": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/FieldValuesForEdit"
                    }
                },
                "File": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/File"
                    }
                },
                "Folder": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/Folder"
                    }
                },
                "LikedByInformation": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/LikedByInformation"
                    }
                },
                "ParentList": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/ParentList"
                    }
                },
                "Properties": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/Properties"
                    }
                },
                "Versions": {
                    "__deferred": {
                        "uri": "https://bertonline.sharepoint.com/sites/prov-1/_api/Web/Lists(guid'7d644d41-d86e-4594-99de-479594a68fd9')/Items(1)/Versions"
                    }
                },
                "HasUniqueRoleAssignments": false,
                "FileSystemObjectType": 0,
                "Id": 1,
                "ServerRedirectedEmbedUri": null,
                "ServerRedirectedEmbedUrl": "",
                "Title": "item updated",
                "FileDirRef": "/sites/prov-1/Lists/FieldTypes",
                "File_x0020_Type": null,
                "ID": 1,
                "Created": "2020-11-10T15:39:12Z",
                "Modified": "2021-10-26T07:16:02Z"
            }
        ]
    }
}
```

**Comment by jansenbe** [comment]
@pxmonti : what you need is possible using REST (see above)...I'll work on implementing this in the coming days.

**Comment by pxmonti** [comment]
Thank you @bjansen, I'll wait for the implementation.

Scarica Outlook per Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jansenbe** [comment]
@pxmonti : I've just pushed a fix for this (see https://github.com/pnp/pnpcore/commit/34687113690375bf87ee5ac62873d50ba9dd3884). With tomorrow's nightly build you should be able to mimic your CSOM scenario. Updated docs will come to https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html

**Comment by pxmonti** [comment]
This is great news! Thank you  @jansenbe !! 
So, if I am understanding correctly, it will be possible to include system fields in the <FieldRef ...... /> section of the CAML query XML definition and they will be extracted too, correct? And for the RoleAssigmnents? Is there anything particular to be done in the method call to fetch these too? Maybe I should simply wait for the documentation to be updated and read these details from there.
Thanks again.

**Comment by jansenbe** [comment]
@pxmonti : yes, just add the system fields as FieldRef's in your CAML query. The code will analyze the CAML XML and ensure the needed fields are added into the $select. Checkout https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html#using-paging-with-loaditemsbycamlquery for a sample. Note the highlighted parts:
![image](https://user-images.githubusercontent.com/7451219/139880900-8f5fc2df-df1d-41b4-acb1-91327e2784c6.png)

**Comment by pxmonti** [comment]
Awesome! I'll install the next nightly build package and test in in my code!
Thank you @jansenbe

**Comment by pxmonti** [comment]
Hi @jansenbe , I've tested the 1.4.34-nightly build and I see that it is 99% working as I was expecting.
I am still unable to fetch the "File_x0020_size" system field, while now I am able to fetch the FileRef field correctly.
I've even tried with item => item.File.Length but it doesn't work.
Here the excerpt of code I am using:

![image](https://user-images.githubusercontent.com/4302704/140034933-75c4c5c6-33c7-4dac-9c27-d7791adb9a0e.png)

What is the correct way of extracting the "File_x0020_Size" field?

**Comment by jansenbe** [comment]
@pxmonti : You've to possible solutions. The red one uses the `File_x0020_Size` property, but you need to get it via also loading the `FieldValuesAsText` property. The green approach loads the `File` with only the `Length` property populated.

![image](https://user-images.githubusercontent.com/7451219/140049921-949a9345-dca0-4a1a-a056-63721f59e544.png)

**Comment by pxmonti** [comment]
Hi @jansenbe , I tested the 2 approaches proposed.
The "red" approach works as a charm and that's enough for my case. A big THANK YOU!
However, just for your information, the "green" approach doesn't work for me because when I try to reference "QueryProperties" method of p.File there's no such method than can be selected and code won't compile.

I am using 1.4.34-nightly build.

**Comment by jansenbe** [comment]
@pxmonti : you need to add a using statement: "using PnP.Core.Model;" before you can use "QueryProperties"

**Comment by jansenbe** [comment]
If all is good, then feel free to close this issue
